socketio/src/client/reconnect.rs

```diff
@@ -0,0 +1,239 @@
+            }
+        }
+        Ok(())
+    }
+
+    fn do_reconnect(&self) -> Result<()> {
+        let new_client = self.builder.clone().connect_manual()?;
+        let mut client = self.client.write()?;
+        *client = new_client;
+
+        Ok(())
+    }
+
+    pub(crate) fn iter(&self) -> Iter {
+        Iter {
+            socket: self.client.clone(),
+        }
+    }
+
+    fn poll_callback(&self) {
+        let mut self_clone = self.clone();
+        // Use thread to consume items in iterator in order to call callbacks
+        std::thread::spawn(move || {
+            // tries to restart a poll cycle whenever a 'normal' error occurs,
+            // it just panics on network errors, in case the poll cycle returned
+            // `Result::Ok`, the server receives a close frame so it's safe to
+            // terminate
+            for packet in self_clone.iter() {
+                if let e @ Err(Error::IncompleteResponseFromEngineIo(_)) = packet {
+                    //TODO: 0.3.X handle errors
+                    //TODO: logging error
+                    let _ = self_clone.disconnect();
+                    break;
+                }
+            }
+            self_clone.reconnect();
+        });
+    }
+}
+
+pub(crate) struct Iter {
+    socket: Arc<RwLock<Client>>,
+}
+
+impl Iterator for Iter {
+    type Item = Result<Packet>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let socket = self.socket.read();
+        match socket {
+            Ok(socket) => match socket.poll() {
+                Err(err) => Some(Err(err)),
+                Ok(Some(packet)) => Some(Ok(packet)),
+                Ok(None) => None,
+            },
+            Err(_) => {
+                // Lock is poisoned, our iterator is useless.
+                None
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::mpsc;
+    use std::thread::sleep;
+
+    use super::*;
+    use crate::error::Result;
+    use crate::{client::TransportType, payload::Payload, ClientBuilder};
+    use bytes::Bytes;
+    use native_tls::TlsConnector;
+    use serde_json::json;
+    use std::time::Duration;
+
+    #[test]
+    fn socket_io_reconnect_integration() -> Result<()> {
+        let url = crate::test::socket_io_restart_server();
+        let connect_count = Arc::new(Mutex::new(0));
+        let close_count = Arc::new(Mutex::new(0));
+        let connect_count_clone = connect_count.clone();
+        let close_count_clone = close_count.clone();
+
+        let socket = ClientBuilder::new(url)
+            .reconnect(true)
+            .max_reconnect_attempts(0) // infinity
+            .reconnect_delay(100, 100)
+            .on(Event::Connect, move |_, socket| {
+                let mut count = connect_count_clone.lock().unwrap();
+                *count += 1;
+            })
+            .on(Event::Close, move |_, socket| {
+                let mut count = close_count_clone.lock().unwrap();
+                *count += 1;
```

Comment:
Did you mean -1? Or not including this at all? Is Close even called/
```suggestion
                *count -= 1;
```

Reply:
I checked the js client, `on('disconnet', () => {})` and  `on('close',, () => {})` both will be called when disconnecting, since we just have `Event::Close`, so `.on(Event::Close, || {})` should be called I think.

Reply:
Just smells odd that both connect and close should increment the same counter both upwards

Reply:
it is not same count，just the lock guard same name，I renamed in new commit